go/pkg/cas/upload.go

```diff
@@ -358,24 +385,67 @@ func (u *uploader) visitDir(ctx context.Context, absPath string) (*repb.Director
 
 	wgChildren.Wait()
 	if subErr != nil {
-		return nil, errors.Wrapf(subErr, "failed to read the directory %q entirely", absPath)
+		return nil, errors.Wrapf(subErr, "failed to read the directory %q entirely", key.AbsPath)
 	}
 
-	item := uploadItemFromDirMsg(absPath, dir)
+	item := uploadItemFromDirMsg(key.AbsPath, dir)
 	if err := u.scheduleCheck(ctx, item); err != nil {
 		return nil, err
 	}
 	return &repb.DirectoryNode{
-		Name:   filepath.Base(absPath),
+		Name:   filepath.Base(key.AbsPath),
 		Digest: item.Digest,
 	}, nil
 }
 
-// visitSymlink converts a symlink to a SymlinkNode and schedules visitation
+// visitSymlink converts a symlink to a directory node and schedules visitation
 // of the target file.
-func (u *uploader) visitSymlink(ctx context.Context, absPath string) (*repb.SymlinkNode, error) {
-	// TODO(nodir): implement
-	panic("not supported")
+// If key.PreserveSymlinks is true, then returns a SymlinkNode, otherwise
+// returns the directory node of the target file.
+func (u *uploader) visitSymlink(ctx context.Context, key fsCacheKey) (interface{}, error) {
```

Comment:
```suggestion
func (u *uploader) visitSymlink(ctx context.Context, key fsCacheKey) (proto.Message, error) {
```

Reply:
this doesn't compile. I could fix those compilation errors by adding more code though if you prefer.

Reply:
I prefer to have nallow scoped types if it can be done with few simple line changes.

Reply:
Done